File: current_session/c++/863.cpp
#include <iostream>
#include <vector>
#include <map>
#include <set>
#include <queue>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
  map<int, vector<int> > connections;
  vector<int> distanceK(TreeNode* root, TreeNode* target, int k) {
    // Build connections first
    buildConnections(NULL, root);

    // BFS
    queue<int> queue;
    queue.push(target->val);
    set<int> visited;
    visited.insert(target->val);
    int dist = 0;
    vector<int> res;
    while (dist<k) {
      int numOfNodesInLayer = queue.size();
      // check the map, pop all nodes in this layer, and push all unvisited nb into queue.
      // for each node in this layer
      while (numOfNodesInLayer--) {
        int cur = queue.front(); queue.pop();
        for (int i = 0; i<connections[cur].size(); i++) {
          int neighbor = connections[cur][i];
          if (visited.find(neighbor) == visited.end()) {
            visited.insert(neighbor);
            queue.push(neighbor);
          }
        }
      }
      dist += 1;
    }
    while (!queue.empty()) {
      res.push_back(queue.front());
      queue.pop();
    }
    return res;
  }
private:
  void buildConnections(TreeNode* parent, TreeNode* child) {
    if (parent!=NULL) {
      connections[parent->val].push_back(child->val);
      connections[child->val].push_back(parent->val);
    }
    if (child->left != nullptr) buildConnections(child, child->left);
    if (child->right != nullptr) buildConnections(child, child->right);
  }
};

File: current_session/python/863.py
"""
https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/
"""
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

import collections
from typing import List

class Solution:
    def distanceK(self, root: TreeNode, target: TreeNode, k: int) -> List[int]:
        # craft the adj list
        self.adj = collections.defaultdict(list)
        def helper(node):
            if node.left:
                self.adj[node].append(node.left)
                self.adj[node.left].append(node)
                helper(node.left)
            if node.right:
                self.adj[node].append(node.right)
                self.adj[node.right].append(node)
                helper(node.right)
            return
        helper(root)

        # DFS
        stack = [(target, 0)]
        visited = set()
        res = []
        while stack:
            x, dist = stack.pop()
            visited.add(x)
            if dist == k:
                res.append(x.val)
                continue
            for nb in self.adj[x]:
                if nb not in visited:
                    stack.append((nb, dist+1))
        return res